{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GSE180286 OBS Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "1. [Introduction](#introduction)\n",
    "2. [Imports](#imports)\n",
    "3. [Path for Assets](#path-for-assets)\n",
    "4. [Data Loading](#data-loading)\n",
    "5. [OBS File Generation](#obs-file-generation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is the start of GSE180286 validation. It is designed to load and process single-cell RNA sequencing data from the GSE180286 dataset. The goal is to prepare the data for further analysis and validation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Necessary imports for the notebook. Please ensure these libraries are installed in your Python environment, \n",
    "# if not then please install them using the requirements.txt file.\n",
    "import glob\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "from gtfparse import read_gtf\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "import gc\n",
    "import numpy as np\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Path for Assets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the relative path of the assets directory, if the assets are in another location please change the path accordingly.\n",
    "assets = \"../assets\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output of this cell is truncated to avoid displaying too many files.\n",
    "# This will list all the files in the GSE180286 directory.\n",
    "gse18_files = glob.glob(f\"{assets}/GSE180286/*\")\n",
    "gsm_files = sorted([f for f in gse18_files if f.endswith(\".txt\")])\n",
    "gsm_file_names = sorted(set(f.split('\\\\')[-1].split('_')[0] for f in gsm_files if f.endswith(\"_matrix.txt\")))\n",
    "# gsm_files, gsm_file_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_adata(gsm_files=gsm_files, gsm_file_names=gsm_file_names):\n",
    "    \"\"\"    Load GSM files into a list of AnnData objects.\n",
    "    Args:\n",
    "        gsm_files (list): List of GSM file names to load.\n",
    "        gsm_file_names (list): List of GSM file names without extensions or the path.\n",
    "    Returns:\n",
    "        list: List of AnnData objects loaded from the provided files.\n",
    "    \"\"\"    \n",
    "    list_data = []\n",
    "    for gsm_file in gsm_files:\n",
    "        path = f\"{gsm_file}\"\n",
    "        print(f\"Loading: {path}\")\n",
    "        \n",
    "        data = sc.read_csv(path, delimiter=\"\\t\").T\n",
    "        \n",
    "        # Convert to sparse\n",
    "        if not issparse(data.X):\n",
    "            data.X = csr_matrix(data.X)\n",
    "        \n",
    "        print(f\"Successfully loaded {gsm_file} with shape {data.shape}\")\n",
    "        list_data.append(data)\n",
    "    adata = ad.concat(list_data, join=\"outer\", label=\"sample\", keys=gsm_file_names)\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading: ../assets/GSE180286\\GSM5457199_A2019-1.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457199_A2019-1.expression_matrix.txt with shape (3267, 25540)\n",
      "Loading: ../assets/GSE180286\\GSM5457200_A2019-2.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457200_A2019-2.expression_matrix.txt with shape (8607, 25501)\n",
      "Loading: ../assets/GSE180286\\GSM5457201_A2019-3.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457201_A2019-3.expression_matrix.txt with shape (16661, 27060)\n",
      "Loading: ../assets/GSE180286\\GSM5457202_B2019-1.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457202_B2019-1.expression_matrix.txt with shape (11467, 32148)\n",
      "Loading: ../assets/GSE180286\\GSM5457203_B2019-2.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457203_B2019-2.expression_matrix.txt with shape (11356, 29607)\n",
      "Loading: ../assets/GSE180286\\GSM5457204_B2019-3.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457204_B2019-3.expression_matrix.txt with shape (11323, 29861)\n",
      "Loading: ../assets/GSE180286\\GSM5457205_C2020-1.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457205_C2020-1.expression_matrix.txt with shape (4161, 30672)\n",
      "Loading: ../assets/GSE180286\\GSM5457206_C2020-2.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457206_C2020-2.expression_matrix.txt with shape (4940, 29843)\n",
      "Loading: ../assets/GSE180286\\GSM5457207_C2020-3.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457207_C2020-3.expression_matrix.txt with shape (8517, 26179)\n",
      "Loading: ../assets/GSE180286\\GSM5457208_D2020-1.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457208_D2020-1.expression_matrix.txt with shape (4064, 29899)\n",
      "Loading: ../assets/GSE180286\\GSM5457209_D2020-2.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457209_D2020-2.expression_matrix.txt with shape (5298, 26256)\n",
      "Loading: ../assets/GSE180286\\GSM5457210_D2020-3.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457210_D2020-3.expression_matrix.txt with shape (8228, 26766)\n",
      "Loading: ../assets/GSE180286\\GSM5457211_E2020-1.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457211_E2020-1.expression_matrix.txt with shape (7521, 30958)\n",
      "Loading: ../assets/GSE180286\\GSM5457212_E2020-2.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457212_E2020-2.expression_matrix.txt with shape (5673, 26816)\n",
      "Loading: ../assets/GSE180286\\GSM5457213_E2020-3.expression_matrix.txt\n",
      "Successfully loaded ../assets/GSE180286\\GSM5457213_E2020-3.expression_matrix.txt with shape (6398, 25668)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "a:\\Capstone Local-GIt\\ghost-cell-busters\\venv\\Lib\\site-packages\\anndata\\_core\\anndata.py:1756: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "# This block takes a really long time to run, so please be patient.\n",
    "# If you are running this locally on a machine please check if you have enough resources available.\n",
    "# If not then please run this on a cloud platform like Google Colab or Kaggle.\n",
    "adata = load_adata(gsm_files=gsm_files, gsm_file_names=gsm_file_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(f\"{assets}/GSE180286/GSE180286_adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGATAAGGGTCA</th>\n",
       "      <td>GSM5457199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGCGTACTG</th>\n",
       "      <td>GSM5457199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAACCTACG</th>\n",
       "      <td>GSM5457199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATAACCTA</th>\n",
       "      <td>GSM5457199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGGTTTGA</th>\n",
       "      <td>GSM5457199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCCAACCAATG</th>\n",
       "      <td>GSM5457213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCCTGAAGCAC</th>\n",
       "      <td>GSM5457213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTACCAGCGGCA</th>\n",
       "      <td>GSM5457213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCTCCTAAAG</th>\n",
       "      <td>GSM5457213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTTTGTAGGG</th>\n",
       "      <td>GSM5457213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117481 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample\n",
       "GGATAAGGGTCA  GSM5457199\n",
       "CCGTGCGTACTG  GSM5457199\n",
       "AGGTAACCTACG  GSM5457199\n",
       "CTGTATAACCTA  GSM5457199\n",
       "AAACAGGTTTGA  GSM5457199\n",
       "...                  ...\n",
       "CTCCAACCAATG  GSM5457213\n",
       "CTCCTGAAGCAC  GSM5457213\n",
       "GTACCAGCGGCA  GSM5457213\n",
       "TACCTCCTAAAG  GSM5457213\n",
       "TGGTTTGTAGGG  GSM5457213\n",
       "\n",
       "[117481 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample\n",
       "GSM5457201    16661\n",
       "GSM5457202    11467\n",
       "GSM5457203    11356\n",
       "GSM5457204    11323\n",
       "GSM5457200     8607\n",
       "GSM5457207     8517\n",
       "GSM5457210     8228\n",
       "GSM5457211     7521\n",
       "GSM5457213     6398\n",
       "GSM5457212     5673\n",
       "GSM5457209     5298\n",
       "GSM5457206     4940\n",
       "GSM5457205     4161\n",
       "GSM5457208     4064\n",
       "GSM5457199     3267\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT genes found: ['MT-ND1', 'MT-ND2', 'MT-ND3', 'MT-ND4', 'MT-ND4L', 'MT-ND5', 'MT-ND6', 'MT-CO1', 'MT-CO2', 'MT-CO3', 'MT-ATP6', 'MT-ATP8', 'MT-CYB', 'MT-RNR1', 'MT-RNR2', 'MT-TP', 'MT-TL1', 'MT-TL2']\n"
     ]
    }
   ],
   "source": [
    "# Common mitochondrial Ensembl gene ID prefixes: ENSG00000198888 → MT-ND1, etc.\n",
    "mt_gene_names = [\n",
    "    \"MT-ND1\", \"MT-ND2\", \"MT-ND3\", \"MT-ND4\", \"MT-ND4L\", \"MT-ND5\", \"MT-ND6\",\n",
    "    \"MT-CO1\", \"MT-CO2\", \"MT-CO3\",\n",
    "    \"MT-ATP6\", \"MT-ATP8\",\n",
    "    \"MT-CYB\", \"MT-RNR1\", \"MT-RNR2\", \"MT-TP\", \"MT-TL1\", \"MT-TL2\"\n",
    "]\n",
    "\n",
    "# Check if these are present\n",
    "mt_present = [gene for gene in mt_gene_names if gene in adata.var_names]\n",
    "print(\"MT genes found:\", mt_present)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'tag', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'havana_transcript', 'exon_number', 'exon_id', 'hgnc_id', 'havana_gene', 'ont', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    }
   ],
   "source": [
    "# Load the GTF file\n",
    "gtf = read_gtf(f\"\"\"{assets}/Gencode/gencode.v44.annotation.gtf\"\"\").to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract gene_id and gene_name\n",
    "genes = gtf[gtf[\"feature\"] == \"gene\"][[\"gene_id\", \"gene_name\"]].drop_duplicates()\n",
    "\n",
    "# Strip version from Ensembl IDs\n",
    "genes[\"gene_id_clean\"] = genes[\"gene_id\"].str.replace(r\"\\..*\", \"\", regex=True)\n",
    "\n",
    "# Create mapping\n",
    "ens_to_symbol = dict(zip(genes[\"gene_id_clean\"], genes[\"gene_name\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map gene symbols in adata_all.var\n",
    "adata.var[\"ensembl_id\"] = adata.var_names.str.replace(r\"\\..*\", \"\", regex=True)\n",
    "adata.var[\"gene_name\"] = adata.var[\"ensembl_id\"].map(ens_to_symbol)\n",
    "\n",
    "# Fill missing symbols with Ensembl ID, then set .var_names\n",
    "# Fill NaNs and ensure all entries are strings\n",
    "adata.var[\"gene_name_clean\"] = (\n",
    "    adata.var[\"gene_name\"].fillna(adata.var[\"ensembl_id\"])\n",
    ").astype(str)\n",
    "\n",
    "# Set .var_names safely\n",
    "adata.var_names = adata.var[\"gene_name_clean\"]\n",
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A1BG',\n",
       " 'A1BG-AS1',\n",
       " 'A1CF',\n",
       " 'A2M',\n",
       " 'A2M-AS1',\n",
       " 'A2ML1',\n",
       " 'A2ML1-AS1',\n",
       " 'A2ML1-AS2',\n",
       " 'A2MP1',\n",
       " 'A3GALT2']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check first few gene names\n",
    "adata.var_names[:10].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify mitochondrial genes by prefix\n",
    "adata.var['mt'] = adata.var_names.str.upper().str.startswith('MT-')\n",
    "\n",
    "# Recalculate QC metrics with updated mitochondrial flags\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], inplace=True)\n",
    "\n",
    "# Assign percent.mito (to match GSE176078 column naming)\n",
    "adata.obs['percent.mito'] = adata.obs['pct_counts_mt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🔍 MT-genes found: ['MT-ATP6', 'MT-ATP8', 'MT-CO1', 'MT-CO2', 'MT-CO3', 'MT-CYB', 'MT-ND1', 'MT-ND2', 'MT-ND3', 'MT-ND4', 'MT-ND4L', 'MT-ND5', 'MT-ND6', 'MT-RNR1', 'MT-RNR2', 'MT-TA', 'MT-TC', 'MT-TD', 'MT-TE', 'MT-TF', 'MT-TG', 'MT-TH', 'MT-TI', 'MT-TK', 'MT-TL1', 'MT-TL2', 'MT-TM', 'MT-TN', 'MT-TP', 'MT-TQ', 'MT-TR', 'MT-TS1', 'MT-TS2', 'MT-TT', 'MT-TV', 'MT-TW', 'MT-TY']\n",
      "MT-total expression: 21227332.0\n",
      "Total expression: 207242224.0\n"
     ]
    }
   ],
   "source": [
    "# Confirm if any MT- genes exist and their total expression\n",
    "mt_genes = adata.var_names[adata.var_names.str.upper().str.startswith(\"MT-\")]\n",
    "print(\"🔍 MT-genes found:\", mt_genes.tolist())\n",
    "\n",
    "# Check summed expression of mitochondrial genes\n",
    "mt_sum = adata[:, mt_genes].X.sum()\n",
    "total_sum = adata.X.sum()\n",
    "print(f\"MT-total expression: {mt_sum}\")\n",
    "print(f\"Total expression: {total_sum}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GGATAAGGGTCA    20.738459\n",
       "CCGTGCGTACTG    18.740520\n",
       "AGGTAACCTACG    18.443533\n",
       "CTGTATAACCTA    17.642475\n",
       "AAACAGGTTTGA    20.140953\n",
       "                  ...    \n",
       "CTCCAACCAATG    12.580645\n",
       "CTCCTGAAGCAC     5.263158\n",
       "GTACCAGCGGCA     6.643356\n",
       "TACCTCCTAAAG    30.769232\n",
       "TGGTTTGTAGGG    18.345324\n",
       "Name: percent.mito, Length: 117481, dtype: float32"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['percent.mito']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene coordinate columns added to adata.var:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name_clean</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58345178.0</td>\n",
       "      <td>58353492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58347718.0</td>\n",
       "      <td>58355455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>chr10</td>\n",
       "      <td>50799409.0</td>\n",
       "      <td>50885675.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9067664.0</td>\n",
       "      <td>9116229.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9065163.0</td>\n",
       "      <td>9068689.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                chromosome       start         end\n",
       "gene_name_clean                                   \n",
       "A1BG                 chr19  58345178.0  58353492.0\n",
       "A1BG-AS1             chr19  58347718.0  58355455.0\n",
       "A1CF                 chr10  50799409.0  50885675.0\n",
       "A2M                  chr12   9067664.0   9116229.0\n",
       "A2M-AS1              chr12   9065163.0   9068689.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse GTF and extract gene features\n",
    "genes = gtf[gtf[\"feature\"] == \"gene\"]\n",
    "gene_pos = genes[[\"gene_name\", \"seqname\", \"start\", \"end\"]].drop_duplicates()\n",
    "gene_pos.columns = [\"gene\", \"chromosome\", \"start\", \"end\"]\n",
    "\n",
    "# Filter gene positions to only those in adata\n",
    "adata.var[\"gene\"] = adata.var_names\n",
    "gene_pos_filtered = gene_pos[gene_pos[\"gene\"].isin(adata.var[\"gene\"])].drop_duplicates(subset=\"gene\")\n",
    "gene_pos_filtered = gene_pos_filtered.set_index(\"gene\")\n",
    "\n",
    "# Remove any existing coordinate columns that would conflict\n",
    "adata.var = adata.var.drop(columns=[\"chromosome\", \"start\", \"end\"], errors=\"ignore\")\n",
    "\n",
    "# Join and reindex\n",
    "adata.var = adata.var.join(gene_pos_filtered, on=\"gene\")\n",
    "adata.var = adata.var.reindex(adata.var_names)\n",
    "\n",
    "print(\"Gene coordinate columns added to adata.var:\")\n",
    "adata.var[[\"chromosome\", \"start\", \"end\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"sample\"].value_counts()\n",
    "adata.obs[\"cnv_reference\"] = \"tumor\"\n",
    "adata.obs.loc[adata.obs[\"sample\"] == \"GSM5457199\", \"cnv_reference\"] = \"normal\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cnv_reference\n",
       "tumor     114214\n",
       "normal      3267\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"cnv_reference\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "percent.mito\n",
       "9.090909     145\n",
       "7.692308     142\n",
       "5.882353     130\n",
       "8.333334     128\n",
       "5.555556     127\n",
       "            ... \n",
       "8.398269       1\n",
       "8.361774       1\n",
       "3.549246       1\n",
       "7.844905       1\n",
       "18.345324      1\n",
       "Name: count, Length: 65191, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"percent.mito\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample chromosome values: ['19', '10', '12', '1', '22', '3', '5', '4', '15', '2']\n",
      "Genes with valid chromosome info: 28623 / 28623\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "a:\\Capstone Local-GIt\\ghost-cell-busters\\venv\\Lib\\site-packages\\anndata\\_core\\anndata.py:1756: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "# Clean chromosome names\n",
    "adata.var[\"chromosome\"] = (\n",
    "    adata.var[\"chromosome\"].astype(str)\n",
    "    .str.replace(\"chr\", \"\", regex=False)\n",
    "    .str.upper()\n",
    ")\n",
    "\n",
    "# Show sample values\n",
    "print(\"Sample chromosome values:\", adata.var[\"chromosome\"].unique().tolist()[:10])\n",
    "\n",
    "# Keep canonical chromosomes only\n",
    "valid_chroms = [str(i) for i in range(1, 23)] + [\"X\", \"Y\"]\n",
    "mask = adata.var[\"chromosome\"].isin(valid_chroms)\n",
    "print(f\"Genes with valid chromosome info: {mask.sum()} / {adata.shape[1]}\")\n",
    "\n",
    "# Subset the full object to valid chromosomes only\n",
    "adata = adata[:, mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome-wise CNV matrix shape: (117481, 24)\n"
     ]
    }
   ],
   "source": [
    "# Clean chromosome info\n",
    "adata.var[\"chromosome\"] = (\n",
    "    adata.var[\"chromosome\"].astype(str)\n",
    "    .str.replace(\"chr\", \"\")\n",
    "    .str.upper()\n",
    ")\n",
    "\n",
    "# Keep only genes with chromosome info\n",
    "valid_mask = adata.var[\"chromosome\"].notna()\n",
    "valid_genes = adata.var[valid_mask]\n",
    "\n",
    "# Convert valid_mask to a NumPy array and get indices\n",
    "valid_idx = np.where(valid_mask.values)[0]\n",
    "\n",
    "# Extract matrix and chromosome info\n",
    "X_cnv = adata.X[:, valid_idx]  # Use .X, not \"X_cnv\"\n",
    "chroms = valid_genes[\"chromosome\"].values\n",
    "\n",
    "# Group gene indices by chromosome\n",
    "chrom_to_idx = defaultdict(list)\n",
    "for idx, chrom in enumerate(chroms):\n",
    "    chrom_to_idx[chrom].append(idx)\n",
    "\n",
    "# Create dataframe for CNV signal\n",
    "cnv_chr_df = pd.DataFrame(index=adata.obs_names)\n",
    "\n",
    "# Compute average signal per chromosome\n",
    "for chrom, indices in chrom_to_idx.items():\n",
    "    X_chr = X_cnv[:, indices]\n",
    "    mean_signal = np.asarray(X_chr.mean(axis=1)).ravel()\n",
    "    cnv_chr_df[chrom] = mean_signal\n",
    "\n",
    "print(\"Chromosome-wise CNV matrix shape:\", cnv_chr_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample CNV scores:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cnv_score</th>\n",
       "      <th>cnv_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGATAAGGGTCA</th>\n",
       "      <td>1.573313</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGCGTACTG</th>\n",
       "      <td>1.022685</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAACCTACG</th>\n",
       "      <td>0.775885</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATAACCTA</th>\n",
       "      <td>0.670172</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGGTTTGA</th>\n",
       "      <td>0.591936</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              cnv_score cnv_reference\n",
       "GGATAAGGGTCA   1.573313        normal\n",
       "CCGTGCGTACTG   1.022685        normal\n",
       "AGGTAACCTACG   0.775885        normal\n",
       "CTGTATAACCTA   0.670172        normal\n",
       "AAACAGGTTTGA   0.591936        normal"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute CNV score as mean absolute deviation across chromosomes\n",
    "cnv_score = cnv_chr_df.abs().mean(axis=1)\n",
    "\n",
    "# Store in AnnData object\n",
    "adata.obs[\"cnv_score\"] = cnv_score\n",
    "\n",
    "# Check result\n",
    "print(\"Sample CNV scores:\")\n",
    "adata.obs[[\"cnv_score\", \"cnv_reference\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['nCount_RNA'] = adata.X.sum(axis=1).A1 if hasattr(adata.X, 'A1') else adata.X.sum(axis=1)\n",
    "mito_genes = adata.var_names.str.startswith('MT-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "ribo_genes = adata.var_names.str.upper().str.startswith(('RPS', 'RPL'))\n",
    "# adata.obs['pct_counts_ribo'] = (adata[:, ribo_genes].X.sum(axis=1).A1 if hasattr(adata.X, 'A1') \n",
    "#                                 else adata[:, ribo_genes].X.sum(axis=1)) / adata.obs['nCount_RNA'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify ribosomal genes\n",
    "ribo_genes = adata.var_names.str.upper().str.startswith(('RPS', 'RPL'))\n",
    "\n",
    "# Calculate ribosomal counts per cell\n",
    "ribo_counts = adata[:, ribo_genes].X.sum(axis=1)\n",
    "if hasattr(ribo_counts, 'A1'):\n",
    "    ribo_counts = ribo_counts.A1\n",
    "else:\n",
    "    ribo_counts = np.array(ribo_counts).ravel()\n",
    "\n",
    "# Calculate percent ribosomal counts per cell\n",
    "adata.obs['pct_counts_ribo'] = ribo_counts / adata.obs['nCount_RNA'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata.obs['percent.mito'] = (adata[:, mito_genes].X.sum(axis=1).A1 if hasattr(adata.X, 'A1') \n",
    "#                               else adata[:, mito_genes].X.sum(axis=1)) / adata.obs['nCount_RNA'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate total counts per cell\n",
    "if hasattr(adata.X, 'A1'):\n",
    "    adata.obs['nCount_RNA'] = adata.X.sum(axis=1).A1\n",
    "else:\n",
    "    adata.obs['nCount_RNA'] = np.array(adata.X.sum(axis=1)).ravel()\n",
    "\n",
    "# Identify mitochondrial genes\n",
    "mito_genes = adata.var_names.str.upper().str.startswith('MT-')\n",
    "\n",
    "# Calculate percent mitochondrial counts per cell\n",
    "mito_counts = adata[:, mito_genes].X.sum(axis=1)\n",
    "if hasattr(mito_counts, 'A1'):\n",
    "    mito_counts = mito_counts.A1\n",
    "else:\n",
    "    mito_counts = np.array(mito_counts).ravel()\n",
    "\n",
    "adata.obs['percent.mito'] = mito_counts / adata.obs['nCount_RNA'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>...</th>\n",
       "      <th>cnv_reference</th>\n",
       "      <th>cnv_score</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>oxphos_score</th>\n",
       "      <th>apoptosis_score</th>\n",
       "      <th>proto_oncogenescore</th>\n",
       "      <th>pct_counts_ribo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGATAAGGGTCA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>7171</td>\n",
       "      <td>8.877940</td>\n",
       "      <td>62319.0</td>\n",
       "      <td>11.040038</td>\n",
       "      <td>34.774306</td>\n",
       "      <td>43.134518</td>\n",
       "      <td>51.916751</td>\n",
       "      <td>63.938767</td>\n",
       "      <td>12924.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.573313</td>\n",
       "      <td>48061.0</td>\n",
       "      <td>-1.038760</td>\n",
       "      <td>-0.072789</td>\n",
       "      <td>G1</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>-0.458551</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>18.690830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGCGTACTG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5783</td>\n",
       "      <td>8.662851</td>\n",
       "      <td>39556.0</td>\n",
       "      <td>10.585498</td>\n",
       "      <td>32.920417</td>\n",
       "      <td>41.404591</td>\n",
       "      <td>51.190717</td>\n",
       "      <td>64.657700</td>\n",
       "      <td>7413.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.022685</td>\n",
       "      <td>31105.0</td>\n",
       "      <td>-0.118217</td>\n",
       "      <td>1.279655</td>\n",
       "      <td>G2M</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>-0.774244</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>17.897444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAACCTACG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5261</td>\n",
       "      <td>8.568266</td>\n",
       "      <td>29734.0</td>\n",
       "      <td>10.300080</td>\n",
       "      <td>31.680904</td>\n",
       "      <td>39.241273</td>\n",
       "      <td>48.627833</td>\n",
       "      <td>62.558014</td>\n",
       "      <td>5484.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.775885</td>\n",
       "      <td>23562.0</td>\n",
       "      <td>-0.062016</td>\n",
       "      <td>1.626887</td>\n",
       "      <td>G2M</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>-0.725756</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>14.043799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATAACCTA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4917</td>\n",
       "      <td>8.500657</td>\n",
       "      <td>25671.0</td>\n",
       "      <td>10.153156</td>\n",
       "      <td>33.345020</td>\n",
       "      <td>41.681275</td>\n",
       "      <td>50.558997</td>\n",
       "      <td>63.986600</td>\n",
       "      <td>4529.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.670172</td>\n",
       "      <td>20544.0</td>\n",
       "      <td>-0.310078</td>\n",
       "      <td>-0.115385</td>\n",
       "      <td>G1</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>1.003910</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>18.589369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGGTTTGA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4784</td>\n",
       "      <td>8.473241</td>\n",
       "      <td>23554.0</td>\n",
       "      <td>10.067094</td>\n",
       "      <td>36.201919</td>\n",
       "      <td>44.293963</td>\n",
       "      <td>52.797826</td>\n",
       "      <td>65.513289</td>\n",
       "      <td>4744.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.591936</td>\n",
       "      <td>18180.0</td>\n",
       "      <td>0.176357</td>\n",
       "      <td>0.726456</td>\n",
       "      <td>G2M</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>0.679093</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>15.891088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample  n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
       "GGATAAGGGTCA  GSM5457199               7171                 8.877940   \n",
       "CCGTGCGTACTG  GSM5457199               5783                 8.662851   \n",
       "AGGTAACCTACG  GSM5457199               5261                 8.568266   \n",
       "CTGTATAACCTA  GSM5457199               4917                 8.500657   \n",
       "AAACAGGTTTGA  GSM5457199               4784                 8.473241   \n",
       "\n",
       "              total_counts  log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "GGATAAGGGTCA       62319.0           11.040038                   34.774306   \n",
       "CCGTGCGTACTG       39556.0           10.585498                   32.920417   \n",
       "AGGTAACCTACG       29734.0           10.300080                   31.680904   \n",
       "CTGTATAACCTA       25671.0           10.153156                   33.345020   \n",
       "AAACAGGTTTGA       23554.0           10.067094                   36.201919   \n",
       "\n",
       "              pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "GGATAAGGGTCA                    43.134518                    51.916751   \n",
       "CCGTGCGTACTG                    41.404591                    51.190717   \n",
       "AGGTAACCTACG                    39.241273                    48.627833   \n",
       "CTGTATAACCTA                    41.681275                    50.558997   \n",
       "AAACAGGTTTGA                    44.293963                    52.797826   \n",
       "\n",
       "              pct_counts_in_top_500_genes  total_counts_mt  ...  \\\n",
       "GGATAAGGGTCA                    63.938767          12924.0  ...   \n",
       "CCGTGCGTACTG                    64.657700           7413.0  ...   \n",
       "AGGTAACCTACG                    62.558014           5484.0  ...   \n",
       "CTGTATAACCTA                    63.986600           4529.0  ...   \n",
       "AAACAGGTTTGA                    65.513289           4744.0  ...   \n",
       "\n",
       "              cnv_reference  cnv_score  nCount_RNA   S_score  G2M_score  \\\n",
       "GGATAAGGGTCA         normal   1.573313     48061.0 -1.038760  -0.072789   \n",
       "CCGTGCGTACTG         normal   1.022685     31105.0 -0.118217   1.279655   \n",
       "AGGTAACCTACG         normal   0.775885     23562.0 -0.062016   1.626887   \n",
       "CTGTATAACCTA         normal   0.670172     20544.0 -0.310078  -0.115385   \n",
       "AAACAGGTTTGA         normal   0.591936     18180.0  0.176357   0.726456   \n",
       "\n",
       "              phase  oxphos_score  apoptosis_score proto_oncogenescore  \\\n",
       "GGATAAGGGTCA     G1     25.207519        -0.458551           25.207519   \n",
       "CCGTGCGTACTG    G2M     19.120603        -0.774244           19.120603   \n",
       "AGGTAACCTACG    G2M     11.705379        -0.725756           11.705379   \n",
       "CTGTATAACCTA     G1      7.692909         1.003910            7.692909   \n",
       "AAACAGGTTTGA    G2M      7.310999         0.679093            7.310999   \n",
       "\n",
       "              pct_counts_ribo  \n",
       "GGATAAGGGTCA        18.690830  \n",
       "CCGTGCGTACTG        17.897444  \n",
       "AGGTAACCTACG        14.043799  \n",
       "CTGTATAACCTA        18.589369  \n",
       "AAACAGGTTTGA        15.891088  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_genes = [\n",
    "    'MCM5', 'PCNA', 'TYMS', 'FEN1', 'MCM2', 'MCM4', 'RRM1', 'UNG', 'GINS2',\n",
    "    'MCM6', 'CDCA7', 'DTL', 'PRIM1', 'UHRF1', 'MCM10', 'HELLS', 'RFC2', 'RPA2',\n",
    "    'NASP', 'RAD51AP1', 'GMNN', 'WDR76', 'SLBP', 'CCNE2', 'UBR7', 'POLD3',\n",
    "    'MSH2', 'ATAD2', 'RAD51', 'RRM2', 'CDC45', 'CDC6', 'EXO1', 'TIPIN', 'DSCC1',\n",
    "    'BLM', 'CASP8AP2', 'USP1', 'CLSPN', 'POLA1', 'CHAF1B', 'BRIP1', 'E2F8'\n",
    "]\n",
    "\n",
    "g2m_genes = [\n",
    "    'HMGB2', 'CDK1', 'NUSAP1', 'UBE2C', 'BIRC5', 'TPX2', 'TOP2A', 'NDC80',\n",
    "    'CKS2', 'NUF2', 'CKS1B', 'MKI67', 'TMPO', 'CENPF', 'TACC3', 'FAM64A',\n",
    "    'SMC4', 'CCNB2', 'CKAP2L', 'CKAP2', 'AURKB', 'BUB1', 'KIF11', 'ANP32E',\n",
    "    'TUBB4B', 'GTSE1', 'KIF20B', 'HJURP', 'CDC20', 'TTK', 'CDC25C', 'KIF2C',\n",
    "    'RANGAP1', 'NCAPD2', 'DLGAP5', 'CDCA3', 'HN1', 'CDC45', 'CDCA8', 'ECT2',\n",
    "    'KIF23', 'HMMR', 'AURKA', 'PSRC1', 'ANLN', 'LBR', 'CKAP5', 'CENPE',\n",
    "    'CTCF', 'NEK2', 'G2E3', 'GAS2L3', 'CBX5', 'CENPA'\n",
    "]\n",
    "\n",
    "oxphos_genes = [\n",
    "    \"ATP5F1A\", \"ATP5F1B\", \"ATP5MC1\", \"ATP5MC2\", \"ATP5ME\", \"ATP5MG\",\n",
    "    \"COX4I1\", \"COX5A\", \"COX6A1\", \"COX6C\", \"NDUFA1\", \"NDUFA2\", \"NDUFA4\",\n",
    "    \"NDUFAB1\", \"NDUFB2\", \"NDUFB3\", \"NDUFS1\", \"NDUFS2\", \"NDUFV1\", \"UQCRC1\",\n",
    "    \"UQCRC2\", \"UQCRH\", \"SDHA\", \"SDHB\", \"SDHC\", \"SDHD\", \"CYCS\"\n",
    "]\n",
    "\n",
    "apoptosis_genes = [\n",
    "    \"BAX\", \"BAK1\", \"BCL2\", \"BCL2L1\", \"BCL2L11\", \"CASP3\", \"CASP6\", \"CASP7\",\n",
    "    \"CASP8\", \"CASP9\", \"TP53\", \"FAS\", \"FASLG\", \"TNFRSF10A\", \"TNFRSF10B\",\n",
    "    \"TNFRSF1A\", \"TNF\", \"AIFM1\", \"APAF1\", \"BAD\", \"BID\", \"CFLAR\", \"DIABLO\",\n",
    "    \"MCL1\", \"NFKB1\", \"NFKBIA\", \"TRADD\", \"XIAP\"\n",
    "]\n",
    "\n",
    "proto_oncogenes = [\n",
    "    'MYC', 'KRAS', 'EGFR', 'BRAF', 'AKT1', 'PIK3CA', 'CCND1', 'ERBB2', 'FGFR1', 'MDM2'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S phase genes found: 43\n",
      "G2M phase genes found: 52\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>...</th>\n",
       "      <th>cnv_reference</th>\n",
       "      <th>cnv_score</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>oxphos_score</th>\n",
       "      <th>apoptosis_score</th>\n",
       "      <th>proto_oncogenescore</th>\n",
       "      <th>pct_counts_ribo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGATAAGGGTCA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>7171</td>\n",
       "      <td>8.877940</td>\n",
       "      <td>62319.0</td>\n",
       "      <td>11.040038</td>\n",
       "      <td>34.774306</td>\n",
       "      <td>43.134518</td>\n",
       "      <td>51.916751</td>\n",
       "      <td>63.938767</td>\n",
       "      <td>12924.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.573313</td>\n",
       "      <td>48061.0</td>\n",
       "      <td>-1.038760</td>\n",
       "      <td>-0.072789</td>\n",
       "      <td>G1</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>-0.458551</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>18.690830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGCGTACTG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5783</td>\n",
       "      <td>8.662851</td>\n",
       "      <td>39556.0</td>\n",
       "      <td>10.585498</td>\n",
       "      <td>32.920417</td>\n",
       "      <td>41.404591</td>\n",
       "      <td>51.190717</td>\n",
       "      <td>64.657700</td>\n",
       "      <td>7413.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.022685</td>\n",
       "      <td>31105.0</td>\n",
       "      <td>-0.118217</td>\n",
       "      <td>1.279655</td>\n",
       "      <td>G2M</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>-0.774244</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>17.897444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAACCTACG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5261</td>\n",
       "      <td>8.568266</td>\n",
       "      <td>29734.0</td>\n",
       "      <td>10.300080</td>\n",
       "      <td>31.680904</td>\n",
       "      <td>39.241273</td>\n",
       "      <td>48.627833</td>\n",
       "      <td>62.558014</td>\n",
       "      <td>5484.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.775885</td>\n",
       "      <td>23562.0</td>\n",
       "      <td>-0.062016</td>\n",
       "      <td>1.626887</td>\n",
       "      <td>G2M</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>-0.725756</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>14.043799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATAACCTA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4917</td>\n",
       "      <td>8.500657</td>\n",
       "      <td>25671.0</td>\n",
       "      <td>10.153156</td>\n",
       "      <td>33.345020</td>\n",
       "      <td>41.681275</td>\n",
       "      <td>50.558997</td>\n",
       "      <td>63.986600</td>\n",
       "      <td>4529.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.670172</td>\n",
       "      <td>20544.0</td>\n",
       "      <td>-0.310078</td>\n",
       "      <td>-0.115385</td>\n",
       "      <td>G1</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>1.003910</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>18.589369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGGTTTGA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4784</td>\n",
       "      <td>8.473241</td>\n",
       "      <td>23554.0</td>\n",
       "      <td>10.067094</td>\n",
       "      <td>36.201919</td>\n",
       "      <td>44.293963</td>\n",
       "      <td>52.797826</td>\n",
       "      <td>65.513289</td>\n",
       "      <td>4744.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.591936</td>\n",
       "      <td>18180.0</td>\n",
       "      <td>0.176357</td>\n",
       "      <td>0.726456</td>\n",
       "      <td>G2M</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>0.679093</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>15.891088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample  n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
       "GGATAAGGGTCA  GSM5457199               7171                 8.877940   \n",
       "CCGTGCGTACTG  GSM5457199               5783                 8.662851   \n",
       "AGGTAACCTACG  GSM5457199               5261                 8.568266   \n",
       "CTGTATAACCTA  GSM5457199               4917                 8.500657   \n",
       "AAACAGGTTTGA  GSM5457199               4784                 8.473241   \n",
       "\n",
       "              total_counts  log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "GGATAAGGGTCA       62319.0           11.040038                   34.774306   \n",
       "CCGTGCGTACTG       39556.0           10.585498                   32.920417   \n",
       "AGGTAACCTACG       29734.0           10.300080                   31.680904   \n",
       "CTGTATAACCTA       25671.0           10.153156                   33.345020   \n",
       "AAACAGGTTTGA       23554.0           10.067094                   36.201919   \n",
       "\n",
       "              pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "GGATAAGGGTCA                    43.134518                    51.916751   \n",
       "CCGTGCGTACTG                    41.404591                    51.190717   \n",
       "AGGTAACCTACG                    39.241273                    48.627833   \n",
       "CTGTATAACCTA                    41.681275                    50.558997   \n",
       "AAACAGGTTTGA                    44.293963                    52.797826   \n",
       "\n",
       "              pct_counts_in_top_500_genes  total_counts_mt  ...  \\\n",
       "GGATAAGGGTCA                    63.938767          12924.0  ...   \n",
       "CCGTGCGTACTG                    64.657700           7413.0  ...   \n",
       "AGGTAACCTACG                    62.558014           5484.0  ...   \n",
       "CTGTATAACCTA                    63.986600           4529.0  ...   \n",
       "AAACAGGTTTGA                    65.513289           4744.0  ...   \n",
       "\n",
       "              cnv_reference  cnv_score  nCount_RNA   S_score  G2M_score  \\\n",
       "GGATAAGGGTCA         normal   1.573313     48061.0 -1.038760  -0.072789   \n",
       "CCGTGCGTACTG         normal   1.022685     31105.0 -0.118217   1.279655   \n",
       "AGGTAACCTACG         normal   0.775885     23562.0 -0.062016   1.626887   \n",
       "CTGTATAACCTA         normal   0.670172     20544.0 -0.310078  -0.115385   \n",
       "AAACAGGTTTGA         normal   0.591936     18180.0  0.176357   0.726456   \n",
       "\n",
       "              phase  oxphos_score  apoptosis_score proto_oncogenescore  \\\n",
       "GGATAAGGGTCA     G1     25.207519        -0.458551           25.207519   \n",
       "CCGTGCGTACTG    G2M     19.120603        -0.774244           19.120603   \n",
       "AGGTAACCTACG    G2M     11.705379        -0.725756           11.705379   \n",
       "CTGTATAACCTA     G1      7.692909         1.003910            7.692909   \n",
       "AAACAGGTTTGA    G2M      7.310999         0.679093            7.310999   \n",
       "\n",
       "              pct_counts_ribo  \n",
       "GGATAAGGGTCA        18.690830  \n",
       "CCGTGCGTACTG        17.897444  \n",
       "AGGTAACCTACG        14.043799  \n",
       "CTGTATAACCTA        18.589369  \n",
       "AAACAGGTTTGA        15.891088  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter genes present in your dataset\n",
    "s_genes_present = [g for g in s_genes if g in adata.var_names]\n",
    "g2m_genes_present = [g for g in g2m_genes if g in adata.var_names]\n",
    "\n",
    "print(f\"S phase genes found: {len(s_genes_present)}\")\n",
    "print(f\"G2M phase genes found: {len(g2m_genes_present)}\")\n",
    "\n",
    "# Score cell cycle\n",
    "sc.tl.score_genes_cell_cycle(adata, s_genes=s_genes_present, g2m_genes=g2m_genes_present)\n",
    "\n",
    "# Confirm it worked\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OxPhos genes used: 27\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    117481.000000\n",
       "mean          0.257516\n",
       "std           0.902100\n",
       "min          -1.720826\n",
       "25%          -0.015261\n",
       "50%           0.050810\n",
       "75%           0.199516\n",
       "max          55.930207\n",
       "Name: oxphos_score, dtype: float64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to genes present in dataset\n",
    "oxphos_genes_present = [g for g in oxphos_genes if g in adata.var_names]\n",
    "\n",
    "# Compute OxPhos score\n",
    "\n",
    "sc.tl.score_genes(adata, gene_list=oxphos_genes_present, score_name=\"oxphos_score\")\n",
    "\n",
    "# Check output\n",
    "print(f\"OxPhos genes used: {len(oxphos_genes_present)}\")\n",
    "adata.obs[\"oxphos_score\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Apoptosis genes found in dataset: 28\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GGATAAGGGTCA   -0.458551\n",
       "CCGTGCGTACTG   -0.774244\n",
       "AGGTAACCTACG   -0.725756\n",
       "CTGTATAACCTA    1.003910\n",
       "AAACAGGTTTGA    0.679093\n",
       "                  ...   \n",
       "CTCCAACCAATG   -0.012774\n",
       "CTCCTGAAGCAC   -0.000782\n",
       "GTACCAGCGGCA   -0.020073\n",
       "TACCTCCTAAAG   -0.007299\n",
       "TGGTTTGTAGGG    0.022941\n",
       "Name: apoptosis_score, Length: 117481, dtype: float64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apoptosis genes\n",
    "apoptosis_genes_present = [g for g in apoptosis_genes if g in adata.var_names]\n",
    "print(f\"Apoptosis genes found in dataset: {len(apoptosis_genes_present)}\")\n",
    "sc.tl.score_genes(adata, gene_list=apoptosis_genes_present, score_name=\"apoptosis_score\")\n",
    "adata.obs[\"apoptosis_score\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Found 10 Proto-oncogenes in dataset: ['MYC', 'KRAS', 'EGFR', 'BRAF', 'AKT1', 'PIK3CA', 'CCND1', 'ERBB2', 'FGFR1', 'MDM2']\n"
     ]
    }
   ],
   "source": [
    "# Score Proto-oncogene activity\n",
    "# Filter to genes in adata\n",
    "valid_protooncogenes = [g for g in proto_oncogenes if g in adata.var_names]\n",
    "sc.tl.score_genes(adata, gene_list=oxphos_genes_present, score_name=\"proto_oncogenescore\")\n",
    "print(f\"Found {len(valid_protooncogenes)} Proto-oncogenes in dataset: {valid_protooncogenes}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>...</th>\n",
       "      <th>cnv_reference</th>\n",
       "      <th>cnv_score</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>oxphos_score</th>\n",
       "      <th>apoptosis_score</th>\n",
       "      <th>proto_oncogenescore</th>\n",
       "      <th>pct_counts_ribo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGATAAGGGTCA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>7171</td>\n",
       "      <td>8.877940</td>\n",
       "      <td>62319.0</td>\n",
       "      <td>11.040038</td>\n",
       "      <td>34.774306</td>\n",
       "      <td>43.134518</td>\n",
       "      <td>51.916751</td>\n",
       "      <td>63.938767</td>\n",
       "      <td>12924.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.573313</td>\n",
       "      <td>48061.0</td>\n",
       "      <td>-1.038760</td>\n",
       "      <td>-0.072789</td>\n",
       "      <td>G1</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>-0.458551</td>\n",
       "      <td>25.207519</td>\n",
       "      <td>18.690830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGCGTACTG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5783</td>\n",
       "      <td>8.662851</td>\n",
       "      <td>39556.0</td>\n",
       "      <td>10.585498</td>\n",
       "      <td>32.920417</td>\n",
       "      <td>41.404591</td>\n",
       "      <td>51.190717</td>\n",
       "      <td>64.657700</td>\n",
       "      <td>7413.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>1.022685</td>\n",
       "      <td>31105.0</td>\n",
       "      <td>-0.118217</td>\n",
       "      <td>1.279655</td>\n",
       "      <td>G2M</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>-0.774244</td>\n",
       "      <td>19.120603</td>\n",
       "      <td>17.897444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTAACCTACG</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>5261</td>\n",
       "      <td>8.568266</td>\n",
       "      <td>29734.0</td>\n",
       "      <td>10.300080</td>\n",
       "      <td>31.680904</td>\n",
       "      <td>39.241273</td>\n",
       "      <td>48.627833</td>\n",
       "      <td>62.558014</td>\n",
       "      <td>5484.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.775885</td>\n",
       "      <td>23562.0</td>\n",
       "      <td>-0.062016</td>\n",
       "      <td>1.626887</td>\n",
       "      <td>G2M</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>-0.725756</td>\n",
       "      <td>11.705379</td>\n",
       "      <td>14.043799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGTATAACCTA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4917</td>\n",
       "      <td>8.500657</td>\n",
       "      <td>25671.0</td>\n",
       "      <td>10.153156</td>\n",
       "      <td>33.345020</td>\n",
       "      <td>41.681275</td>\n",
       "      <td>50.558997</td>\n",
       "      <td>63.986600</td>\n",
       "      <td>4529.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.670172</td>\n",
       "      <td>20544.0</td>\n",
       "      <td>-0.310078</td>\n",
       "      <td>-0.115385</td>\n",
       "      <td>G1</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>1.003910</td>\n",
       "      <td>7.692909</td>\n",
       "      <td>18.589369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGGTTTGA</th>\n",
       "      <td>GSM5457199</td>\n",
       "      <td>4784</td>\n",
       "      <td>8.473241</td>\n",
       "      <td>23554.0</td>\n",
       "      <td>10.067094</td>\n",
       "      <td>36.201919</td>\n",
       "      <td>44.293963</td>\n",
       "      <td>52.797826</td>\n",
       "      <td>65.513289</td>\n",
       "      <td>4744.0</td>\n",
       "      <td>...</td>\n",
       "      <td>normal</td>\n",
       "      <td>0.591936</td>\n",
       "      <td>18180.0</td>\n",
       "      <td>0.176357</td>\n",
       "      <td>0.726456</td>\n",
       "      <td>G2M</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>0.679093</td>\n",
       "      <td>7.310999</td>\n",
       "      <td>15.891088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample  n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
       "GGATAAGGGTCA  GSM5457199               7171                 8.877940   \n",
       "CCGTGCGTACTG  GSM5457199               5783                 8.662851   \n",
       "AGGTAACCTACG  GSM5457199               5261                 8.568266   \n",
       "CTGTATAACCTA  GSM5457199               4917                 8.500657   \n",
       "AAACAGGTTTGA  GSM5457199               4784                 8.473241   \n",
       "\n",
       "              total_counts  log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "GGATAAGGGTCA       62319.0           11.040038                   34.774306   \n",
       "CCGTGCGTACTG       39556.0           10.585498                   32.920417   \n",
       "AGGTAACCTACG       29734.0           10.300080                   31.680904   \n",
       "CTGTATAACCTA       25671.0           10.153156                   33.345020   \n",
       "AAACAGGTTTGA       23554.0           10.067094                   36.201919   \n",
       "\n",
       "              pct_counts_in_top_100_genes  pct_counts_in_top_200_genes  \\\n",
       "GGATAAGGGTCA                    43.134518                    51.916751   \n",
       "CCGTGCGTACTG                    41.404591                    51.190717   \n",
       "AGGTAACCTACG                    39.241273                    48.627833   \n",
       "CTGTATAACCTA                    41.681275                    50.558997   \n",
       "AAACAGGTTTGA                    44.293963                    52.797826   \n",
       "\n",
       "              pct_counts_in_top_500_genes  total_counts_mt  ...  \\\n",
       "GGATAAGGGTCA                    63.938767          12924.0  ...   \n",
       "CCGTGCGTACTG                    64.657700           7413.0  ...   \n",
       "AGGTAACCTACG                    62.558014           5484.0  ...   \n",
       "CTGTATAACCTA                    63.986600           4529.0  ...   \n",
       "AAACAGGTTTGA                    65.513289           4744.0  ...   \n",
       "\n",
       "              cnv_reference  cnv_score  nCount_RNA   S_score  G2M_score  \\\n",
       "GGATAAGGGTCA         normal   1.573313     48061.0 -1.038760  -0.072789   \n",
       "CCGTGCGTACTG         normal   1.022685     31105.0 -0.118217   1.279655   \n",
       "AGGTAACCTACG         normal   0.775885     23562.0 -0.062016   1.626887   \n",
       "CTGTATAACCTA         normal   0.670172     20544.0 -0.310078  -0.115385   \n",
       "AAACAGGTTTGA         normal   0.591936     18180.0  0.176357   0.726456   \n",
       "\n",
       "              phase  oxphos_score  apoptosis_score proto_oncogenescore  \\\n",
       "GGATAAGGGTCA     G1     25.207519        -0.458551           25.207519   \n",
       "CCGTGCGTACTG    G2M     19.120603        -0.774244           19.120603   \n",
       "AGGTAACCTACG    G2M     11.705379        -0.725756           11.705379   \n",
       "CTGTATAACCTA     G1      7.692909         1.003910            7.692909   \n",
       "AAACAGGTTTGA    G2M      7.310999         0.679093            7.310999   \n",
       "\n",
       "              pct_counts_ribo  \n",
       "GGATAAGGGTCA        18.690830  \n",
       "CCGTGCGTACTG        17.897444  \n",
       "AGGTAACCTACG        14.043799  \n",
       "CTGTATAACCTA        18.589369  \n",
       "AAACAGGTTTGA        15.891088  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample',\n",
       " 'n_genes_by_counts',\n",
       " 'log1p_n_genes_by_counts',\n",
       " 'total_counts',\n",
       " 'log1p_total_counts',\n",
       " 'pct_counts_in_top_50_genes',\n",
       " 'pct_counts_in_top_100_genes',\n",
       " 'pct_counts_in_top_200_genes',\n",
       " 'pct_counts_in_top_500_genes',\n",
       " 'total_counts_mt',\n",
       " 'log1p_total_counts_mt',\n",
       " 'pct_counts_mt',\n",
       " 'percent.mito',\n",
       " 'cnv_reference',\n",
       " 'cnv_score',\n",
       " 'nCount_RNA',\n",
       " 'S_score',\n",
       " 'G2M_score',\n",
       " 'phase',\n",
       " 'oxphos_score',\n",
       " 'apoptosis_score',\n",
       " 'proto_oncogenescore',\n",
       " 'pct_counts_ribo']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OBS File Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the .obs DataFrame to CSV\n",
    "adata.obs.to_csv(f\"{assets}/GSE180286/GSE180286_obs.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
